### DISCUSSION TITLE
Return types of state variables with nested ref/computed properties don't match the specified type

### DISCUSSION THREAD

**Discussion by DanRiess** [vuejs/pinia#2696] [Help and Questions] [answered]
### Reproduction

https://play.pinia.vuejs.org/#eNqVVFFP2zAQ/ismL22lNgExMalqEWxiEtO0TcBbw0NILsWQnC3bKUVV//vOdpImpWPjpY3Pd999332XbIJLKcNVBcE0mOlUcWmYBlNJViS4nMeB0XFwHmMUsfQR0meWigoNqNBolgvFYJ2UsoAYZ5Gvplw6GKBoYoBOs6hzCMYBL6VQZlImMnzSAqnvJkbG4vqC2k2Zi9iY5MgTG4mDyD2HoMvJgxIvmjg8UfK4Sb0gEVEGKyNEoSeJ5L7s0Ripp1GUZkj5GRR8pUIEE6Esozc1F2fhWXgSFfwhokYRxwzW/TZUMcmg/B/0JvXiODz5FH52qKvTGrS0qBZ0G+OWxmJ0KjDny72hpKKUvAD1SxousD+cpCjEy3cXM6qClqIz6kD8Sa89698KaHwr6MgyiVqC8ddXtz9hTc/tZSmyijx+7/IGtCgqy9GnfakwI9qdPMf22nnMcXmnr9YGUDeiLFE3DZfvxvz1Hek7uqfhaWeKu+2kEfqFYhuWQc4Rbo1QMGba/t2JG8g127JciZIN3GoNYmwrbO/KQEZZY+Z+FORtOpGj5BjNqwR2TevN5p4aJiVMqYMihfYsFU8p0EGbYVU+gKJXgRjXCD+4tiNpQHjWhUjpxVkK9dqNcWqpp67z4r5GgrWjTkukDau0V0uQHe3DQT2ewZgNR2x+7vv5ksKTIFjLsqa0uD8nCJI+XNyPWKL37mxfxvIKU+sQ06KEb3iZpkCO4fKyThzW0NNG6Kjx0XYWBYSFaJNGtZk9YLK6LtXEuy5W9JVS2PLu1NU3m/aOTD9MbdyFdjtE/WP0E3GLQvKbYQ47d37akLW1c58eNi0ZfTBRGNY12Hrly3fUrK3tPu6YNHg+PnTPox7mvhMemOSQv29Y9QdIMHseH1llPvVvJu4rXhzTRhASKCXUv4oPaAxXSVHBh1D64voA4vko2P4Bvrdf2Q==

### Steps to reproduce the bug

This issue might be related to #2658, but the linked PR does not solve it. 

If a state variable contains a prop that is itself a ref or a computed value, the return type of that variable will 'simplify' that away (i.e. a prop with type ComputedRef<number> will become just number).

That means, however, that using the return type of this state variable as a parameter in a function will throw a ts error.

A workaround is to expose the state variable through a getter in the store, but I kinda hate that because it breaks consistency. And since the state variable has to be exported to keep reactivity, there is no way of knowing if I can access it directly or have to use the getter without looking into the store, where it may or may not be documented.

``` ts
type Item = {
  name: string
  price: ComputedRef<number>
}

type Listing = {
  id: string
  category: string
  items: Item[]
}

export const useStore = defineStore('counter', () => {
  const listings: Ref<Listing[]> = ref([])

  function someFnAcceptingAListing(listing: Listing) {
    console.log(listing)
  }

  function getListings() {
    return listings
  }

  return { listings, someFnAcceptingAListing, getListings }
})



const store = useStore()

// Access listings in different ways.
const exportedListings = store.listings // ❌ not type Listing[]
const { listings: storeToRefsListings } = storeToRefs(store) // ❌ not type Ref<Listing[]>
const getterListings = store.getListings() // ✅ Ref<Listing[]> !

// Try to pass listings to a function that should accept them as parameter.
// All of these work fine in runtime, but some the first 2 throw a ts error.
store.someFnAcceptingAListing(exportedListings[0]) // ❌ error
store.someFnAcceptingAListing(storeToRefsListings.value[0]) // ❌ error
store.someFnAcceptingAListing(getterListings.value[0]) // ✅ ok!
```

### Expected behavior

All 3 methods of accessing the state variable should result in the same return type.

### Actual behavior

I suppose returning the state variable directly triggers some internal unwrapping, which breaks the type? Anyways, returning such a complex typed variable changes the original type.

**Comment by posva** [comment, marked as answer]
The problem comes the casting on the ref being wrong. [refs auto unwrap nested refs](https://vuejs.org/guide/essentials/reactivity-fundamentals.html#additional-ref-unwrapping-details) (and computed, etc). Once the ref is properly typed, [type errors go away](https://play.pinia.vuejs.org/#eNqVVFFP2zAQ/ismL22lNAYxMalqK9jEJKaJTcCemj6E5FIMiW3ZTinq+t93tpM0gY5tL21yvvvuu+8+ZxtcSBmtKwgmwVSniklDNJhKkiLhq1kcGB0H85hTStIHSJ9IKipuQEVGk1woApuklAXEnJVSKEO2lYZbIxTsSK5ESQYR3VcMYh7zKfVtEBRfDGB5YgDfprTzEoSBRxyXiYweteBIcBtzQuL6AHlNiIvYmGScJTYSB9Q9R6DL8b0SzxpbP2Jy2KSe47Q0g7URotDjRDJf9mCM1BNK04xjfgYFW6uIg6FclvRNzflZdBad0ILdU2xEGc9g02+DFeMMyn9Bb1LPj6OTD9FHh7o+rUFLi2pBdzHfoSxGp4LnbPVKlFSUkhWgvkvDBO+LkxSFeP7qYkZV0FJ0Gz0Qf9Qbz/qHApRvDZ2xTKJWYPzx5e01bPC5PSxFVqEZ3ju8AS2KynL0aZ8qniHtTp5je+V2zPjqTl9uDHDdDGWJOjVcvpP58zuj7+meRqcdFfemRAkb75IMcsa9f0Oi7d+duIFck8bNzlpo47bC9q4MZJgVEvfzkz+rRF6D9lEdEgV5C4B03S0wLxLIFRqezDxZnpQwwZ4KZ7bvUrEUAx38Ka/Ke1B4OXCGGuEb01akBoRlXYgUr9JKqJdujGFLPXGdF8saCTZuGLSVNqS5vwjZUWM4qAUbhGQ4IrO57+dLCk9CYwmOOq05LZbz4WI5sg0IySue2uUQLUr4wi/SFHBZfHVRJw9rjEk70a83QjbA81GzXtteFBAVogXAfm7HvabogLpWI/m6WOFXTvGWfKeuPtm2Z+iFw7TDLrSzlpvXy+L8g5o0ig47Z15yyNramU+PWi3xg8uFId0t24X58j01u9vWpnsmDZ6PD93zqIdpDbXfVAOM4+CS37DqC4gw/WpyZCfzqX9a8OuJF8dLhwRKCfW34gMzRuukqOC/UPrD9QHE01Gw+w3YRHi2)

Note that putting a computed/ref within the nested data can create issues during SSR because it allows the data source to be duplicated during serialization.

**Reply by DanRiess** [reply]
I see, thanks for the reply! I wasn't aware of the UnwrapNestedRefs type.

Do you happen to know if there is a recommended best practice on how to handle nested computeds? 
I did it because I have to derive data from 3 different sources that I wanted to keep updating individually, but still bundle the information for easier access.

**Reply by posva** [reply]
There isn't a _recommended_ way because it really depends on what you want to do. Sharing refs is allowed but it does mean that you will need to use `UnwrapRefs` and some `// @ts-expect-error` until TS allows having getters and setters with different types for arbitrary keys (there is an issue about this in both Vue and TS)
